src/Framework/Framework/Compilation/Binding/ExpressionBuildingVisitor.cs

```diff
@@ -527,6 +527,19 @@ protected override Expression VisitFormattedExpression(FormattedBindingParserNod
 
         protected override Expression VisitVoid(VoidBindingParserNode node) => Expression.Default(typeof(void));
 
+        protected override Expression VisitArrayInitializer(ArrayInitializerExpression node)
+        {
+            var initializers = node.ElementInitializers.Select(e => Visit(e));
+
+            var firstInitializer = initializers.FirstOrDefault();
+
+            var firstElementType = firstInitializer?.Type ?? typeof(object);
+
+            var arrayElementType = initializers.All(i => i.Type.IsAssignableFrom(firstElementType)) ? firstElementType : typeof(object);
```

Comment:
I'd prefer to throw an exception if we can't find a good element type. We can always add more handled cases later without breaking anyone. And in unhandled cases it would be better to support C# cast operation instead of doing magic implicit conversions

`[(object) A, (object) B]`


or a cast method like this: `Cast<object>(A)`

Reply:
Added exceptions